std/complex.d

```diff
@@ -1047,3 +1047,294 @@ Complex!T sqrt(T)(Complex!T z)  @safe pure nothrow @nogc
     // Test ease of use (vanilla toString() should be supported)
     assert(complex(1.2, 3.4).toString() == "1.2+3.4i");
 }
+
+/**
+ * Calculates e$(SUPERSCRIPT x).
+ * Params:
+ *      x = A complex number
+ * Returns:
+ *      The complex base e exponential of `x`
+ */
+Complex!T exp(T)(Complex!T x) @trusted pure nothrow @nogc // TODO: @safe
+{
+    static import std.math;
+
+    // Handle special cases explicitly here, as fromPolar will otherwise
+    // cause them to return Complex!T(NaN, NaN), or with the wrong sign.
+    if (std.math.isInfinity(x.re))
+    {
+        if (std.math.isNaN(x.im))
+        {
+            if (std.math.signbit(x.re))
+                return Complex!T(0, std.math.copysign(0, x.im));
+            else
+                return x;
+        }
+        if (std.math.isInfinity(x.im))
+        {
+            if (std.math.signbit(x.re))
+                return Complex!T(0, std.math.copysign(0, x.im));
+            else
+                return Complex!T(T.infinity, -T.nan);
+        }
+        if (x.im == 0.0)
+        {
+            if (std.math.signbit(x.re))
+                return Complex!T(0.0);
+            else
+                return Complex!T(T.infinity);
+        }
+    }
+    if (std.math.isNaN(x.re))
+    {
+        if (std.math.isNaN(x.im) || std.math.isInfinity(x.im))
+            return Complex!T(T.nan, T.nan);
+        if (x.im == 0.0)
+            return x;
+    }
+    if (x.re == 0.0)
+    {
+        if (std.math.isNaN(x.im) || std.math.isInfinity(x.im))
+            return Complex!T(T.nan, T.nan);
+        else
+            return Complex!T(1.0, 0.0);
```

Comment:
there is still a phase you need to worry about here.

Reply:
Forgot the `if (x.im == 0.0)` bit. :-)

Reply:
Right, and fixing this should now cause https://github.com/dlang/phobos/pull/7456#discussion_r418970871 to fail.